### ISSUE TITLE
Readable error messages

### ISSUE THREAD

**Issue by Morwenn** [Morwenn/cpp-sort#28]
In its current state, the library produces unreadble error messages an can spawn several hundreds of lines of cryptic error messages for a simple missing `const` without a single hint to what may be wrong. Honestly, this is a major drawback and the crazy amount of SFINAE insanity doesn't help.

Concepts will probably help to reduce the error madness, but we can't put static assertions all over the place since most of the interesting features heavily rely on SFINAE.

PLZ HAPL!!

**Comment by vendethiel** [comment]
(unreadable*. here, did that help?)

**Comment by Morwenn** [comment]
@vendethiel Haha, it was more along the lines of « we want readable error messages », but you're right, I will change the title :p

**Comment by vendethiel** [comment]
:D sorry, I'm terrible at C++. Just ask my coworkers ;)

**Comment by Morwenn** [comment]
Apparently, [some tricks exist][1] to display better error messages upon SFINAE failure. I can probably check whether some of them could be used to improve **cpp-sort**'s error messages too.


  [1]: https://github.com/SuperV1234/cppnow2016/blob/master/trip_report.md#inline-static-error

**Comment by vendethiel** [comment]
have you looked at the sqlpp guy's talk?

**Comment by Morwenn** [comment]
Yes I have (well, I have read the slides). The « strong `typedef` via inheritance » as well as the mention of concepts come from here. It's even possible that I opened this issue because of the talk actually.

**Comment by vendethiel** [comment]
that makes sense 👍

**Comment by Morwenn** [comment]
[Yet another article](https://gracicot.github.io/tricks/2017/07/01/deleted-function-diagnostic.html) that gives ideas to get better error messages with SFINAE.

**Comment by Morwenn** [comment]
Two small changes that could reduce the compiler call stack in error messages:
* Replace use of `std::distance(it1, it2)` and `std::advance(it, n)` with `(it2 - it1)` and `it += n` when the algorithm is known to deal with random-access iterators.
* Don't use `std::begin` and `std::end` when the type of the container is known and unlikely to change.

Those little functions are used everywhere in the code base, don't offer much outside of sufficiently generic contexts and make can make calls stacks grow. Moreover it's easier to track the origin of a problem directly when `(it2 - it1)` appears directly in the algorithm rather than in the implementation of `std::distance`.
